### PR TITLE
AwsSdk2Transport implementation

### PR DESCRIPTION
**Pull request by mtimmerm**
This PR adds a transport implementation that can connect to AWS OpenSearch service using IAM authentication.  Http clients and request signers from the AWS SDK v2 are used directly.

In order to test with the integration tests, you should set system property tests.awsSdk2support.domainHost to the FQDN of an OpenSearch service domain, and  tests.awsSdk2support.domainRegion to its region if it's not us-east-1.  It must accept IAM authentication using your configured default credentials.  You can set the AWS_PROFILE environment variable if necessary.

The AwsSdk2Transport supports both synchronous and asynchronous requests with either synchronous or asynchronous AWS HTTP clients, as well as automatic request and response compression and bulk requests.

This PR also adds a builder for TransportOptions, which has always been missing since RestClientTransport was the only transport.  

An optional dependency on the required AWS SDK packages has been added.

### CODE REVIEW

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java:185]: Maybe we can refactor this out into a private method that handles errors? Split the bottom into `performRequestSync`?

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java:325]: We need to extract the signing region into a parameter here.

**Review by dblock — CHANGES_REQUESTED** [review]
First, thank you! Amazing.

I did a quick review to get started and left some relatively small comments. I'll take a more thorough look as we go.

How does this interact with wanting compressed/chunked requests?

@reta you might be interested in looking at this too

**Comment by reta** [comment]
@mtimmerm @dblock I think we should not bundle anything vendor-specific (like AWS fe) into `java-client`, with that being said,  it may not be worth creating separate repository but `java-client-aws` or `java-client-transport-aws` would make sense, wdyt?

File: java-client/build.gradle.kts
Comment (mtimmerm) [reply java-client/build.gradle.kts:126]: There's a tests.rest.cluster property to set the cluster used in the rest client integration tests.  I was going for consistency with that, but you can have whatever property you want.

File: java-client/build.gradle.kts
Comment (mtimmerm) [reply java-client/build.gradle.kts:163]: It depends on the feature registration above.  I don't think we should call the feature "awsSdkV2", but "awsSdk2support" would be OK

File: java-client/src/main/java/org/opensearch/client/transport/aws/AsyncByteArrayContentPublisher.java
Comment (mtimmerm) [reply java-client/src/main/java/org/opensearch/client/transport/aws/AsyncByteArrayContentPublisher.java:9]: I tried that originally, but checkstyle complained.

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java
Comment (VachaShah) [inline java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java:373]: Should we log the error here as well?

**Review by VachaShah — COMMENTED** [review]
Thank you very much @mtimmerm!

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java
Comment (mtimmerm) [reply java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java:158]: Well, I can make a constructor that allows you to specify both, and then these two can delegate there.  Couldn't hurt.

File: java-client/src/main/java/org/opensearch/client/transport/aws/AsyncCapturingSubscriber.java
Comment (mtimmerm) [reply java-client/src/main/java/org/opensearch/client/transport/aws/AsyncCapturingSubscriber.java:86]: Sure

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java
Comment (mtimmerm) [reply java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java:185]: This exception handling is specific to the case where an async client is used to service a synchronous request, so I'd like to leave it there.  I can split out the bottom of the two `performRequest...` methods into private ones and remove the mutual recursion, though.

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java
Comment (mtimmerm) [reply java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java:325]: Oops. yes, of course

File: java-client/src/test/java/org/opensearch/client/opensearch/integTest/awsv2/AwsV2SearchTest.java
Comment (mtimmerm) [reply java-client/src/test/java/org/opensearch/client/opensearch/integTest/awsv2/AwsV2SearchTest.java:32]: Sure

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java
Comment (mtimmerm) [reply java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java:373]: I don't think so.  It'll get logged where it gets caught

File: java-client/build.gradle.kts
Comment (mtimmerm) [reply java-client/build.gradle.kts:126]: changing to tests.awsSdk2support.domainHost and tests.awsSdk2support.domainRegion

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java
Comment (mtimmerm) [reply java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java:9]: Changing to AwsSdk2Transport

File: java-client/src/test/java/org/opensearch/client/opensearch/integTest/awsv2/AwsV2SearchTest.java
Comment (mtimmerm) [reply java-client/src/test/java/org/opensearch/client/opensearch/integTest/awsv2/AwsV2SearchTest.java:32]: used package `aws` for consistency with source, and `AwsSdk2SearchIT` with `AwsSdk2TransportTestCase` base

**Comment by mtimmerm** [comment]
@reta About the idea of publishing a separate package out of the same repo:

I don't really have a strong opinion on that, but I think that is something the code owners would do after this is merged.  I would certainly not be offended.

My not-so-strong opinion, though, is that this library could really use a few transport implementations (this one, a java.net.http one, and maybe an apache client one), and it already includes a transport that is dependent on the RestClient.  As long as the rest client one is in here, it's consistent to have the others here.  Moving the RestClient transport out will require a major version bump.

Also, it is harmless, since the dependencies are optional, and the gradle "feature variant" functionality was designed to accomplish exactly this sort of thing:  https://blog.gradle.org/optional-dependencies

**Comment by mtimmerm** [comment]
Switched to [DRAFT] for a bit.  I'm going to add a specific transport options class with compression params.

**Comment by dblock** [comment]
> @mtimmerm @dblock I think we should not bundle anything vendor-specific (like AWS fe) into `java-client`, with that being said, it may not be worth creating separate repository but new `java-client-aws` or `java-client-transport-aws` module would make sense, wdyt?

Is the only difference with optional dependencies (gradle features) that we tell users to include `software.amazon.awssdk .auth` in their dependencies themselves vs. a hypothetical `java-client-aws` which includes it as its own dependency? If so, feels like making RestClient an optional dependency is the best option. It also solves the fact that opensearch-java is dependent on including a specific version of OpenSearch REST client, we might want to boot that altogether in favor of the apache transport client.

File: java-client/src/main/java/org/opensearch/client/transport/aws/AsyncByteArrayContentPublisher.java
Comment (mtimmerm) [reply java-client/src/main/java/org/opensearch/client/transport/aws/AsyncByteArrayContentPublisher.java:9]: @dblock CheckStyle is configured to require all 3 headers in all source files (see header.java.txt).  This means that you cannot create any files of your own without the ElasticSearch license.

I expect this is not what you want going forward, but I don't think it's my place to mess with that.  I don't mind licensing my contributions to ElasticSearch as well, even though they don't appear in any ES repos, so I will add all 3 headers to make checkstyle happy.

File: java-client/src/main/java/org/opensearch/client/transport/aws/AsyncCapturingResponseHandler.java
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/transport/aws/AsyncCapturingResponseHandler.java:9]: Same as above.

**Review by dblock — CHANGES_REQUESTED** [review]
Sorry I missed the update @mtimmerm.

There are a bunch of files with more than just SPDX headers still, like java-client/src/main/java/org/opensearch/client/util/OpenSearchRequestBodyBuffer.java. I just want to confirm that these are copied from other files that have that license header? If they are brand new, please remove all but the SPDX header. 

I'm good with this code! I'd want @reta to give another pair of eyes for any must have's.

How does one use this? I think our README should really have some basic examples, maybe it's worth starting with this? I'll try to write some, so not an ask for this PR unless you feel super motivated.

**Comment by mtimmerm** [comment]
Most of these files are new, but you have checkstyle configured to ensure that all 3 headers are present.  I don't mind having them all the headers in this code, but if you want just the SPDX one then I can configure your checkstyle to skip the directories I put stuff in (your integration test directory is similarly skipped already).  Please advise.

The integration test demonstrates how to use it: https://github.com/mtimmerm/opensearch-java/blob/7ba29d1069e1f0daaf8219e8c57d0ed34d92ba9c/java-client/src/test/java/org/opensearch/client/opensearch/integTest/aws/AwsSdk2TransportTestCase.java#L90

**Comment by dblock** [comment]
> Most of these files are new, but you have checkstyle configured to ensure that all 3 headers are present. I don't mind having them all the headers in this code, but if you want just the SPDX one then I can configure your checkstyle to skip the directories I put stuff in (your integration test directory is similarly skipped already). Please advise.
> 
> The integration test demonstrates how to use it: https://github.com/mtimmerm/opensearch-java/blob/7ba29d1069e1f0daaf8219e8c57d0ed34d92ba9c/java-client/src/test/java/org/opensearch/client/opensearch/integTest/aws/AwsSdk2TransportTestCase.java#L90

The rule is that we *only add* SPDX headers for all new code, and that we *don't remove* any existing copyright. I think checkstyle should just have SPDX rather than exclude directories? We can enforce non-removal in PRs I think.

**Comment by mtimmerm** [comment]
OK, I'll update the license headers when #178 is merged.

**Comment by dblock** [comment]
Thanks for hanging in with me here @mtimmerm. I talked to the right people and confirmed https://github.com/opensearch-project/opensearch-java/pull/178, which is now merged. Care to update?

File: java-client/src/main/java/org/opensearch/client/transport/aws/AsyncCapturingSubscriber.java
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/transport/aws/AsyncCapturingSubscriber.java:62]: typo: parameter

**Review by dblock — APPROVED** [review]
I'll merge this and we can cleanup comments and other minor stuff as we go. I'll take on experimenting with this and writing some samples/docs.

**Review by VachaShah — APPROVED** [review]
LGTM, thank you @mtimmerm!

**Comment by VachaShah** [comment]
> I'll merge this and we can cleanup comments and other minor stuff as we go. I'll take on experimenting with this and writing some samples/docs.

Merging this according to the comment.

**Comment by dblock** [comment]
I have a working demo in https://github.com/dblock/opensearch-java-client-demo/tree/opensearch-java-2.1.0. If you see anything that can be improved in that demo, please speak up, since I'm going to start incorporating some of this into opensearch/aws docs.

**Comment by mtimmerm** [comment]
That looks good to me.  Maybe add a commented out config for a credential provider.  Most of my use cases require that.

**Comment by dblock** [comment]
> That looks good to me. Maybe add a commented out config for a credential provider. Most of my use cases require that.

What does it look like in your usage? Feel free to just PR into my repo, too.

**Comment by dblock** [comment]
Btw, this was released in 2.1.0 🥳

**Comment by mtimmerm** [comment]
@dblock That's awesome!  Thanks for the note.  I will be so happy to remove the bespoke version from my product.

**Pull request merged**
